Introduce template formatter

mwmiller opened the pull request:
This handles, in some way, everything which the parser does.  We could probably use some tests, documentation and improved configurability.

mwmiller commented:
@bartblast This is not yet fully ready for merge.  It could use a number of improvements, certainly as pertain to documentation but maybe also correctness.  I wanted to create the PR, though, so that you could provide any feedback on how those might proceed.  I'm trying to keep things in line with the code conventions I see around.  Let me know where I have failed on that front as well.

mwmiller commented:
> * `~HOLO""` causes some heartburn.  The sigil is ineterpeted as `~HOLO"` with a text `"` in the body.

This was the same as the problem which followed it.  I was injecting a newline into an empty template... of course.

bartblast commented:
Hey @mwmiller, awesome work, thanks! :)

## Initial thoughts:
* I like the state machine approach - that's a solid foundation for this kind of parsing work
* Are there things we should add/update in the parser itself to make formatting easier? For example, adding metadata to parser output that could help the formatter?
* You mentioned "not sure about the best way to apply testing without making a mess of your repo" - I'm not sure I understand why tests would make a mess. Can you clarify what you're concerned about?
* Would love to see some basic tests to understand how it's working in practice

mwmiller commented:
`Inspect.Algebra` doesn't feel like it is going to get me where I want to be.  I feel like I have to over-specify for the fairly simple needs I have.

I ran the current version over my Hologram project and identified some more failings.  I will develop some tests related thereto and iterate further tomorrow.

mwmiller commented:
> **Scope for v1:** I think for the first version it could be a good idea to just focus on indentation and attributes/props formatting. Once we've got that working on solid foundations, we can build on it incrementally.

Sure thing.  It is working well-enough for my personal project today.  The resulting HTML subjectively "looks better" and "doesn't break."  Those are my two criteria for this project, so that's sweet!

> **Expressions:** I wouldn't touch expressions at all for now - you're right to leave whitespace untouched there since we can't be sure about significance.

Sure thing.  I do mangle some whitespace in `:text` nodes, but I leave `:expressions` alone.  I **do** have to touch them in one place.  The conditional expression in the `Parser` output of an `if` block is enclosed in `{}`.  These don't exist in the actual source, so I strip the out in what I consider a [dumb way](https://github.com/bartblast/hologram/pull/250/files#diff-028b0c7ec80f60645ce33437dabbf3644bceba0acb2ed48df6970d7e00c62eafR171-R177).

> **Configuration:** To be honest, I'm not sure we need configuration options. That's one of the pros of `mix format` - it stopped pointless discussions about conventions and left it to the community. What do you think about keeping it opinionated?

I'm in favor.  Even when a formatter author makes the "wrong" choice it saves me a bunch of puttering around in making my own choice.  So this is the answer for which I was hoping.  The formatter does get the options from `.formatter.exs` if there are any we ought to respect.  This project has an extensively modified `.formatter.exs`. I poked through and didn't see anything about which I should concern myself (especially since I leave expressions untouched.)

> **Self-closing tags:** I think we should go with `<MyComponent />` (space before slash) - that seems to be the standard convention.

Cool.  This is already in there, then.  I might also remove the `holo_indent` configuration  option (default: `"  "`).  Two space indentation seems pretty standard in the community.

> **Block/inline elements:** I don't know of any authoritative list of block/inline tags, but if there is one, it's probably somewhere in the [HTML spec](https://html.spec.whatwg.org/).

I ended up using a list from MDN.  It has to do with what **renders** as block/inline in a browser.  It turns out this is a decent enough proxy for how you want your source structured as well.  I added and removed a couple tags based on visual review.

> You can also take a look at the [Phoenix LiveView formatter]

I have looked that code over a bit (particularly while on my `Inspect.Algebra` goose chasing. Their approach to blocking is approximately the opposite of mine. They keep a list of `inline` elements which shouldn't break; I keep a list of `block` elements which should. I (perhaps obviously) think my approach is better, though they are both kinda "meh" and fragile.  (On a callback note, they allow configuring the formatter with a different set of  `inline` elements which inspired some earlier questions.)

> I haven't dug deep into your code yet, so this is just brainstorming at this point. What are your thoughts on the DOMTree approach?

This was my original approach, actually.  It had some problems; perhaps some were self-inflicted.  

Probably the most notable one was that "templates don't have to be HTML documents."  They don't even have to be HTML tree fragments.  It's possible for any given template to be "unrooted", coming out of the parser with a (non-whitespace) `:text` node.  I spent some time wrapping in a fake root and treating it like a tree.  It felt pretty hacky

As you mentioned, there is also the issue of, say `{%if ...}` conditionals.  Is that contained block in the tree?  What if there is an `{else}`?  Are those two siblings in the same tree, even though they will never be rendered together? You mention treating it as a single node but I can imagine complex structures wherein this gets even messier.

For source formatting purposes, I can (mostly) ignore the HTML rendering and just worry about the source rendering.  Once I try to put it into a "working" DOMTree, I have problems with "validating" that tree.

bartblast commented:
> > **Expressions:** I wouldn't touch expressions at all for now - you're right to leave whitespace untouched there since we can't be sure about significance.
> 
> Sure thing. I do mangle some whitespace in `:text` nodes, but I leave `:expressions` alone. I **do** have to touch them in one place. The conditional expression in the `Parser` output of an `if` block is enclosed in `{}`. These don't exist in the actual source, so I strip the out in what I consider a [dumb way](https://github.com/bartblast/hologram/pull/250/files#diff-028b0c7ec80f60645ce33437dabbf3644bceba0acb2ed48df6970d7e00c62eafR171-R177).

Do you mean leaving this untouched: `<MyWidget extra={ @nonsense } plus={@more} />`?
Actually, we should trim the whitespace inside the outer curly braces, because they're treated as regular tuples by the Hologram compiler with special behavior depending on what they contain. They're tuples in essence, so they should be formatted as tuples.

---

> I might also remove the `holo_indent` configuration option (default: `" "`). Two space indentation seems pretty standard in the community.

👍 

---

> Probably the most notable one was that "templates don't have to be HTML documents." They don't even have to be HTML tree fragments. It's possible for any given template to be "unrooted", coming out of the parser with a (non-whitespace) `:text` node. I spent some time wrapping in a fake root and treating it like a tree. It felt pretty hacky

Yeah, templates can be just a list of nodes - this is actually more flexible than LiveView's requirement for a single root node, which often feels artificial. I'm curious about your concern with the "virtual" root node approach though. What specifically makes it feel hacky to you? Is it the temporary nature of it, or something about how it affects the formatting logic?

---

> As you mentioned, there is also the issue of, say `{%if ...}` conditionals. Is that contained block in the tree? What if there is an `{else}`? Are those two siblings in the same tree, even though they will never be rendered together? You mention treating it as a single node but I can imagine complex structures wherein this gets even messier.

When I mentioned "DOM tree", I meant our document object model tree, not HTML specifically. Thinking about it this way, any structure that evaluates to an expression can be treated as a node in the tree. Looking at how Elixir code is generated from templates, it's just a tree structure anyway. So for `if` blocks, we could have something like `%IfBlock{do: ..., else: ...}` (or any other pair like true/false, then/else, etc.).

---

> I am wondering if "control blocks" (like `{%if ...}`) should also be "source blocks" where the contained structure is set off with line breaks and indentation. My first instinct is "of course, it's right in the name!", but I also can see where it might be something like:
> 
> ```
> <h3>Hello, {%if @user}{@user}{%else}guest{/if}</h3>
> ```
> 
> I would personally write this actual example something more like {@user || "guest"}, but it could be more complex with potential `nil`s or something. I doubt if anyone would want that to suddenly become:
> 
> ```
> <h3>Hello, 
>   {%if @user}
>     {@user}
>   {%else}
>     guest
>   {/if}</h3>
> ```

I don't think we should format control blocks like `{%if ...}`. You correctly noticed that it could introduce unintended whitespace inside the branches, but the developer wrote it as a single line intentionally. We should respect their formatting choice and only format in cases when it improves readability without changing the developer's intent.

mwmiller commented:
I just pushed a new single commit which greatly simplifies things.  I don't have time to look into some of the new edges introduced.  I also haven't taken care of the expression squeezing just yet.  I expect to have time to work on that tomorrow or the next day.

mwmiller commented:
I added some tests and made some more progress.  I also realized I may have broken some semantics with my whitespace trimming.  I don't have time to fix it right now, so I will make that my first task on my next go 'round.

mwmiller commented:
This now does mostly what I want.  I need to look further into divining developer intent around conditional blocks.

It also fails rather spectacularly on `lib/hologram/ui/runtime.ex`.  This file has a templated `<script>` tag which gets mangled and becomes uncompilable.  I tried adding an `inviolable` tracker for `<script>` (and potentially other future) tags.  It works but produces poor output.  So I'm going to let that percolate.

Over the rest of the repo it changes 17 other files.  Some in ways with which I agree and others in ways which I find "acceptable"... but might mask developer intent.

So, I'll be back at this in a few days when I have time and a hopeful epiphany.

mwmiller commented:
I had some more time with this today.  I tried (again) to create an `inviolable` section when something is wrapped in `<script>` tags.  However, looking more closely at the `Parser` output for `lib/hologram/ui/runtime.ex` it doesn't contain all of the source information.

I can certainly skip this file in the formatter, but I want to make sure it is an intentional `HOLO` special-case.  I still have lots to work out for making it friendly, so this isn't blocking.

mwmiller commented:
It seems like it might be only the `{%raw}` blocks missing from the parser.  A minimal example from the repo (`test/features/app/pages/template_syntax/raw_block_page.ex`):

```
 ~HOLO"{%raw}{%if false}abc{@var}xyz{/if}{/raw}"
```

Parses to:
```
[text: "{%if false}abc{@var}xyz{/if}"]
```

I haven't looked too deeply, but it looks like this is to allow for a macro-esque building of a  `HOLO` template inside a `HOLO` template.  This doesn't confront me much either way, but the loss of information from the parse makes it essentially impossible to reproduce the source directly.

I'm going to poke further at the Parser to see if there are other subtleties I have missed.

bartblast commented:
> `raw` is the issue which makes this wholly untenable at this point. I see two ways to proceed. Either have the parser include the `raw` block tags or do a static inspection of the source before passing it to the parser. In this second case, I would simply skip formatting of template sources. I could add documentation or a warning.

This shouldn't be a problem - feel free to add the raw block tags similar to how other blocks are flagged in the parser output. Currently `:raw?` is just an opaque internal flag in the parser context, but you can make it explicit by outputting block tags.
Looking at the parser, you'll see that `{%if}`, `{%for}`, and `{%else}` all produce explicit tags in the output:
* `{:block_start, "else"}`
* `{:block_start, {"for", expression}}`
* `{:block_end, "for"}`

Right now `{%raw}` and `{/raw}` (lines 646-658 in parser.ex) only call `enable_raw_mode()` and `disable_raw_mode()` which toggle the internal `:raw?` flag, but they don't emit any block tags. You could probably follow the same pattern as if/for blocks - something like this (just an idea from the top of my head, haven't verified it, there could be more to this):

```elixir
def parse_tokens(%{raw?: false} = context, :text, [{:symbol, "{%raw}"} = token | rest]) do
  context
  |> maybe_add_text_tag()
  |> add_processed_token(token)
  |> add_processed_tag({:block_start, "raw"})  # Add this
  |> enable_raw_mode()
  |> parse_tokens(:text, rest)
end

def parse_tokens(%{raw?: true} = context, :text, [{:symbol, "{/raw}"} = token | rest]) do
  context
  |> maybe_add_text_tag()
  |> add_processed_token(token)
  |> add_processed_tag({:block_end, "raw"})  # Add this
  |> disable_raw_mode()
  |> parse_tokens(:text, rest)
end
```
